bite-size-notes/week-6-a-CTCS_SCP.md

```diff
@@ -0,0 +1,171 @@
+
+$$
+g^j(t, x(t), u(t)) = 1 - \left\| S^j(r(t) - r^j_{\text{obs}})\right\| \leq 0
+$$
+ 
+given the shape operator $S^j$ of the obstacle $j$-th.
+
+Discretizing the previous problem leads, for each node $i$ and obstacle $j$, to 
+
+$$
+g^j(t_i, x_i, u_i) = 1 - \left\| S^j(r_i - r^j_{\text{obs}})\right\| \leq 0
+$$
+
+What happens between node $i$ and node $i+1$ with the discrete formulation? The result follows: drone, at the first obstacle, goes :boom:. The SCP solver updates the nodes (black dots) in such manner they are not inside of the visible ellipses, i.e. the obstacles. However, SCP is not detecting that **between** the node points, drone is piercing through the obstacles. In other words, for SCP everything is fine, and drone is safe. No way.
+
+![CTCS_noCS](figs/CTCS_noCS.gif)
+*Retrieved with permission from [4]
+
+### CTCS-granting Optimal Control
+To overcome the inter-sample constraint violation problem, [2] makes use of *exterior penalty functions*, thus allowing a similar approach to the dynamics shooting; consider the fictitious state $y \in \mathbb{R}^{g+h}$ with time derivative
+
+$$
+\dot{y}(t) \coloneqq 1^\top|g(t, x(t), u(t))|^2_+ + 1^\top h(t, x(t), u(t))^2
+$$
+
+ From $y$ dynamics definition, $\dot{y}(t) \geq 0\; \forall t \in [t_0, t_f]$; if boundary condition $y_0 = y_f$ is satisfied, then $\dot{y}(t) = 0 \;  \forall t \in [t_0, t_f]$, implying constraints $g(t, x(t), u(t)) \leq 0, h(t, x(t), u(t))=0$ are satisfied $\forall t \in [t_0, t_f]$. Therefore, the discretized problem reads
+
+$$\begin{array}{rl}
+\underset{x_i,p_i,t_f}{\text{minimize}} & L(t_{N+1}, x_{N+1}) \\
+\text{subject to} & x_{i+1} = x_{i} + \int_{t_{i}}^{t_{i+1}} f(\tau, x(\tau), \xi_i(\tau, p_i))\text{d}\tau  \quad i = 1\dots, N \\
+& y_{i+1} = y_{i} + \int_{t_{i}}^{t_{i+1}} [1^\top |g(\tau, x(\tau), \xi_i(\tau, p_i))|^2_+ + 1^\top h(\tau, x(\tau), \xi_i(\tau, p_i))^2]\text{d}\tau  \quad i = 1\dots, N \\
+& y_{i+1} - y_{i} \leq \varepsilon \quad i = 1\dots, N \\
+& t_1 = t_0, t_{N+1} = t_f \\
+& Q(t_1, x_1, t_{N+1}, x_{N+1}) = 0 \\
+\end{array}
+$$
+
+where the positive variable $\varepsilon$ allows to 1) grant Linear Independence Constraint Qualification (LICQ), making the problem tractable, and 2) leverage an exact penalty approach [2]. In addition, the given $\varepsilon$ is linked with the maximum pointwise violation of the considered path constraints, according to the following relation 
+
+$$
+\int_{t_k}^{t_{k+1}} |g^j(\tau, x(\tau), \xi_i(\tau, p_i)) + \delta_{g^j}(\varepsilon)|^2_+ \text{d}t \leq \varepsilon \Rightarrow g^j(\tau, x(\tau), \xi_i(\tau, p_i)) \leq 0, \; \forall t \in [t_k, t_{k+1}]
+$$
+
+where the monotonic increasing function $\delta_{g^j}(\varepsilon)$ can be estimated at a first iteration; the SCP approach can then be adopted, for decreasing $\varepsilon$, up to satisfaction of a desired tolerance $\texttt{tol}$: $\delta_{g^j}(\varepsilon) < \text{\texttt{tol}}$.
+
```

Comment:
Adding a line on the role of the monotonic increasing function would be helpful, as it wasn't stated before.

Reply:
Commend addressed